corehq/apps/es/case_search.py

```diff
@@ -297,12 +297,12 @@ def case_property_range_query(case_property_name, gt=None, gte=None, lt=None, lt
     except ValueError:
         pass
 
-    # if its a date, use it
+    # if its a date or datetime, use it
     # date range
-    parse = parse_datetime if is_user_input else parse_date
     kwargs = {
-        key: parse(value) for key, value in kwargs.items()
-        if value is not None and parse(value) is not None
+        key: parse_date(value) if parse_date(value) else parse_datetime(value)
+        for key, value in kwargs.items()
+        if value is not None and (parse_date(value) or parse_datetime(value))
```

Comment:
A couple things about this feel weird to me (neither of which is new in this PR, I recognize).

First, this doesn't support passing in date objects, only string representations of dates, which are then turned into date objects.  That's odd, and forces us to `isoformat()` dates into strings just to have them parsed back into dates or datetimes.  This ultimately resolves to [this function](https://github.com/dimagi/commcare-hq/blob/a0cbc781ea8b00f6d054e4e12873da3d120829bf/corehq/apps/es/filters.py#L76-L76) which can handle strings, dates, or datetimes.  In my opinion, this should support those same three options.  It may be simpler to split `case_property_range_query` into a `numeric_range` and `date_and_time_range` version.

Secondly is this conditional:
```if value is not None and (parse_date(value) or parse_datetime(value))```
Range functions accept gt, gte, lt, and lte args, and in all likelihood, two are three are going to be `None`, so it makes sense to ignore those, but what does it mean to ignore those that can't be parsed as a date or datetime?  If someone calls
```python
case_property_range_query('last_modified', gte='20023-07-24', lt='2023-07-25')
```
That is an error (`20023` is not a valid year).  Instead, this will just ignore the `gte` clause and return all cases modified prior to `2023-07-25`

Reply:
Good point about supporting date and datetime objects - [ef25fc3](https://github.com/dimagi/commcare-hq/pull/33286/commits/ef25fc3f636072a70c809ab0f2b9d52ce53792b9)

I agree with your second point too, I added better handling for both when `parse_date` and `parase_datetime` throws a value error and if they are both None.
https://github.com/dimagi/commcare-hq/pull/33286/commits/0a10e0071e2e451b9c048f29ade117bdcfe299f0
https://github.com/dimagi/commcare-hq/pull/33286/commits/a9eb6b6787eff8227e6a56a0011a16d3a193b2c0